tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/DecisionTreeTrainer.scala

```diff
@@ -306,17 +309,37 @@ class DecisionTreeTrainer(
     }
   }
 
-  private def computeInformationGainUnoptimized(
+  /** Computes the standard entropy-based information gain metric for decision tree growing.
+    *
+    * @param data a source of feature vectors
+    * @param featureVectorSubset the indices of the feature vectors that interest us
+    * @param featureSubset the indices of the features that interest us
+    * @return a sequence of pairs (FEATURE, GAIN), where GAIN is FEATURE's information gain
+    */
+  private def computeEntropyBasedInformationGain(
     data: FeatureVectorSource,
-    featureVectorSubset: Seq[Int], featureSubset: Seq[Int]
+    featureVectorSubset: Seq[Int],
+    featureSubset: Seq[Int]
   ): Seq[(Int, Double)] = {
 
-    require(featureVectorSubset.nonEmpty)
-    require(featureSubset.nonEmpty)
+    def computeOutcomeEntropy(featureVectorSubstream: Seq[FeatureVector]) = {
+      val outcomeHistogram: Map[Int, Int] =
+        featureVectorSubstream groupBy { _.outcome.get } mapValues { _.size }
+      computeEntropy(outcomeHistogram)
+    }
 
-    val featureVectorSubstream: Seq[FeatureVector] = featureVectorSubset map {
-      data.getNthVector(_)
+    def computeEntropy(histogram: Map[Int, Int]) = {
```

Comment:
Inline above? Seems weird to have it be a function in the same scope with the same nested level . . .